src/software/estop/UARTCommunication.cpp

```diff
@@ -0,0 +1,67 @@
+
+#include "UARTCommunication.h"
+
+#include <boost/asio.hpp>
+#include <g3log/loglevels.hpp>
+#include <g3log/g3log.hpp>
+#include <iostream>
+#include <boost/function.hpp>
+#include <boost/exception/all.hpp>
+
+UARTCommunication::UARTCommunication(){}
+
+
+int UARTCommunication::openPort(io_service &ioService, int baudRate, std::string port)
+{
+    int uartCharacterSize = 8;
+    try
+    {
+        serial_ = serial_port_ptr(std::make_shared<boost::asio::serial_port>(boost::asio::serial_port(ioService, port)));
+
+        serial_->set_option(boost::asio::serial_port_base::baud_rate(baudRate));
+        serial_->set_option(boost::asio::serial_port::flow_control(
+                boost::asio::serial_port::flow_control::none));
+        serial_->set_option(
+                boost::asio::serial_port::parity(boost::asio::serial_port::parity::none));
+        serial_->set_option(
+                boost::asio::serial_port::stop_bits(boost::asio::serial_port::stop_bits::one));
+        serial_->set_option(boost::asio::serial_port::character_size(
+                boost::asio::serial_port::character_size(uartCharacterSize)));
+    }
+    catch(boost::exception const&  ex)
+    {
+        LOG(WARNING) << "Unable to setup serial communication with port: " << port;
+        LOG(WARNING) << "Boost exception: " <<diagnostic_information(ex);
+        return -1;
+    }
+
+    return 0;
+}
+
+int UARTCommunication::serialRead(unsigned char *read_val, int num_bytes, boost::system::error_code& ec) {
+    if(serial_->is_open()){
+        long unsigned int read_size = boost::asio::read(*serial_, boost::asio::buffer(read_val, num_bytes), ec);
+        return (int) read_size;
+    } else{
+        return -1;
```

Comment:
there are 3 scenarios i can think of where we would not expect `serialRead/Write` to work and could throw an exception:

1) we call `closePort` successfully, then we call `serialRead/write`, currently we return `-1`
2) if `openPort` was called successfully, then arduino was unplugged, we attempt to read but fail and get `eof` error code in `ec`
3) if arduino was unplugged from start, and we try to call `openPort`, we display LOG(warning) and any subsequent call to `serialRead/write` will segfault

It would make sense to throw an exception for all these cases so its clear there is an error. To handle the exceptions, is there anything more we can do other than print log warnings/errors?

We could also make use of the error code `ec` that is being passed in.

Reply:
If we can't open the port, we should fail out with a nice error message, segfaults might be hard to track down. 

So if the port doesn't open `LOG(FATAL) << "e-stop not connected, but real robot detected on network, please connect an e-stop and relaunch"` would be nice. 

You can check the return value in the e-stop handler (or whatever instantiates this connection) and print this there, so returning a bool here seems reasonable imo. The serial connection either works or it doesn't, there is no recovery action we can take on any exception

Reply:
made changes so that: 
-if `openPort` fails, `false` is returned (with the expectation being that the caller handles this error) 
- if serial port is closed or disconnected, and `serialRead/Write` is called, exception is thrown